### PR TITLE
Factorize code to allow generic models + add decoder version

### PR DESCRIPTION
**Pull request by thomasw21**

### CODE REVIEW

File: evaluation/data_collator.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+from dataclasses import dataclass
```
Comment (VictorSanh) [inline evaluation/data_collator.py:1]: nit: could you copy-paste the header (important to give back credit since we are basing this code on existing code)

File: evaluation/model.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    def forward(self, batch):
+        batch_size, prefix_length = batch["input_ids"].shape
+        model_inputs = {
+            "input_ids": torch.cat([batch["input_ids"], batch["labels"]], dim=-1),
+            "attention_mask": torch.cat([batch["attention_mask"], batch["labels_attention_mask"]], dim=-1),
+        }
+        # Set position ids correctly to take care of padding tokens between inputs_ids and labels
+        position_ids = torch.cumsum(model_inputs["attention_mask"].to(torch.long), dim=-1)
+        model_inputs["position_ids"] = position_ids
```
Comment (VictorSanh) [inline evaluation/model.py:91]: shouldn't that start from 0 rather than 1?

**Review by VictorSanh — APPROVED** [review]
thanks @thomasw21 , lgtm!
i had to scratch my head with these padding handling but I think this is correct

Double-checking: the arguments in L266-281 (in particular `add_special_tokens=False`)were specifically adapted for t zero. depending on the training is done, they might not be valid anymore for another architecture... wanted to make sure there is no confusion here

File: evaluation/data_collator.py (hunk 1 above)
Comment (thomasw21) [reply evaluation/data_collator.py]: I'm unclear on that statement. Does git not track automatically those (move/copies)?
